### PR TITLE
James 1717 Handle HTML in vacation responses

### PR DESCRIPTION
**Pull request by chibenwa**
- Store the htmlBody for vacation
 - Use it to right emails
 - Sent emails should have a plain text version extracted from HTML when no plain text is provided.

### CODE REVIEW

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/mailet/VacationReply.java [hunk 1]
```diff
@@ -57,33 +63,76 @@ public Builder receivedMailRecipient(MailAddress mailRecipient) {
...
+                Multipart multipart = new MimeMultipart("mixed");
+                addPlainPart(multipart, vacation.getTextBody());
+                addHtmlPart(multipart, vacation.getHtmlBody());
+                return multipart;
+            } catch (IOException e) {
+                throw new MessagingException("Cannot read specified content", e);
+            }
+        }
+
+        private Multipart addPlainPart(Multipart multipart, Optional<String> textOptional) throws MessagingException, IOException {
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/mailet/VacationReply.java]: textOptional is a strange variable name, no? (and later htmlOptional)

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/mailet/VacationReply.java [hunk 2]
```diff
@@ -57,33 +63,76 @@ public Builder receivedMailRecipient(MailAddress mailRecipient) {
...
+            if (textOptional.isPresent()) {
+                return addTextPart(multipart, textOptional.get(), "text/plain");
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/mailet/VacationReply.java]: can be done with a lamba, no?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/mailet/VacationReply.java [hunk 3]
```diff
@@ -57,33 +63,76 @@ public Builder receivedMailRecipient(MailAddress mailRecipient) {
...
+            }
+            return multipart;
+        }
+
+        private Multipart addHtmlPart(Multipart multipart, Optional<String> htmlOptional) throws MessagingException, IOException {
+            if (htmlOptional.isPresent()) {
+                return addTextPart(multipart, htmlOptional.get(), "text/html");
+            }
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/mailet/VacationReply.java]: idem

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/mailet/VacationReply.java (hunk 1 above)
Comment (chibenwa) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/mailet/VacationReply.java]: Ok

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/mailet/VacationReply.java (hunk 2 above)
Comment (chibenwa) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/mailet/VacationReply.java]: Yes. Updated

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/mailet/VacationReply.java (hunk 3 above)
Comment (chibenwa) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/mailet/VacationReply.java]: Updated

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/vacation/Vacation.java [hunk 4]
```diff
@@ -58,7 +59,7 @@ public Builder toDate(Optional<ZonedDateTime> toDate) {
             return this;
         }
 
-        public Builder textBody(String textBody) {
+        public Builder textBody(Optional<String> textBody) {
             this.textBody = textBody;
```
Comment (mbaechler) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/vacation/Vacation.java]: checkNotNull

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/vacation/Vacation.java [hunk 5]
```diff
@@ -78,23 +84,26 @@ public Builder copy(Vacation vacation) {
         }
 
         public Vacation build() {
-            Preconditions.checkNotNull(textBody);
-            return new Vacation(isEnabled.orElse(DEFAULT_DISABLED), fromDate, toDate, textBody, subject);
+            Preconditions.checkState(textBody.isPresent() || htmlBody.isPresent() || !isEnabled.orElse(DEFAULT_DISABLED), "textBody or htmlBody property of vacationResponse object should not be null when enabled");
```
Comment (mbaechler) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/vacation/Vacation.java]: you should first compute the `isEnable` then have a `if` then check `textBody || htmlBody`

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/vacation/Vacation.java [hunk 6]
```diff
@@ -58,7 +59,7 @@ public Builder toDate(Optional<ZonedDateTime> toDate) {
             return this;
         }
 
-        public Builder textBody(String textBody) {
+        public Builder textBody(Optional<String> textBody) {
```
Comment (mbaechler) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/vacation/Vacation.java]: you should also provide the same method with `String textBody` to easy the call when we know we have some content.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/VacationResponse.java [hunk 7]
```diff
@@ -44,11 +44,12 @@ public static Builder builder() {
     @JsonPOJOBuilder(withPrefix = "")
     public static class Builder {
         private String id;
-        private boolean isEnabled;
+        private boolean isEnabled = false;
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/VacationResponse.java]: should be Optional<Boolean> to handle default value in `build()`

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/mailet/VacationReply.java [hunk 8]
```diff
@@ -57,33 +63,76 @@ public Builder receivedMailRecipient(MailAddress mailRecipient) {
...
+            } else {
+                reply.setContent(generateMultipart());
+            }
             return reply;
         }
+
+
+        private Multipart generateMultipart() throws MessagingException {
+            try {
+                Multipart multipart = new MimeMultipart("mixed");
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/mailet/VacationReply.java]: no constant for that  ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/VacationIntegrationTest.java [hunk 9]
```diff
@@ -122,6 +123,35 @@ public void jmapVacationShouldGenerateAReplyWhenActive() throws Exception {
     }
 
     @Test
+    public void jmapVacationShouldHaveSupportForHtmlMail() throws Exception {
+        /* Test scenario :
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/VacationIntegrationTest.java]: The test says it already

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/vacation/Vacation.java (hunk 6 above)
Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/vacation/Vacation.java]: Done

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/VacationResponse.java (hunk 7 above)
Comment (chibenwa) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/VacationResponse.java]: Done

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/mailet/VacationReply.java (hunk 8 above)
Comment (chibenwa) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/mailet/VacationReply.java]: Extracted, as long as other values here

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/VacationIntegrationTest.java (hunk 9 above)
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/VacationIntegrationTest.java]: Removed

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/vacation/Vacation.java (hunk 5 above)
Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/vacation/Vacation.java]: Ok

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/vacation/Vacation.java (hunk 4 above)
Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/vacation/Vacation.java]: Added. As long to other methods here.

**Comment by rouazana** [comment]
:+1:

**Comment by chibenwa** [comment]
test this please

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MimeMessageBodyGenerator.java [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+import com.github.fge.lambdas.Throwing;
+import com.google.common.base.Preconditions;
+
+public class MimeMessageBodyGenerator {
+    public static final String MIXED = "mixed";
+
+    private final HtmlTextExtractor htmlTextExtractor;
+
+    @Inject
+    public MimeMessageBodyGenerator(HtmlTextExtractor htmlTextExtractor) {
```
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MimeMessageBodyGenerator.java]: @VisibleForTesting

Comment (chibenwa) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MimeMessageBodyGenerator.java]: Ok

**Comment by aduprat** [comment]
:+1:

**Pull request closed without merging**